File: CS210/HR_Lab4.py
import math

def closest_prime(n):
    diff = n
    primes = sieve(2*n)
    for i in range(len(primes)):
        if diff <= abs(n - primes[i]):
            return primes[i - 1]
        else:
            diff = abs(n - primes[i])


def sieve(n):
    primes = [True] * n
    for p in range(3, int(math.sqrt(n)) + 1, 2):
        if p**2 > n:
            break
        elif primes[p]:
            for i in range(p*p, n, 2*p):
                primes[i] = False
    return [2] + [p for p in range(3, n, 2) if primes[p]]


#---------------------------------------------#

n = int(input())
print(closest_prime(n))

File: CS210/HR_Lab4.java
import java.util.Scanner;

public class HR_Lab4 {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = Integer.parseInt(in.nextLine());

        System.out.printf("%d\n", closest(n, sieve(2*n)));
    }

    static int[] sieve(int n) {
        boolean[] p = new boolean[n];
        p[0] = true; p[1] = true;
        int c = 2;
        for (int i = 2; i*i < n; i++) {
            if (!p[i]) {
                for (int j = 2*i; j < n; j += i) {
                    if (!p[j]) {
                        p[j] = true; c++;
                    }
                }
            }
        }

        int[] primes = new int[n - c];
        for (int i = 0, j = 0; i < n; i++) {
            if (!p[i]) {
                primes[j++] = i;
            }
        }
        return primes;
    }

    static int closest(int n, int[] ns) {
        int min = Integer.MAX_VALUE;
        int closest = n;
        for (int i : ns) {
            if (Math.abs(n - i) < min) {
                min = Math.abs(n - i);
                closest = i;
            }
            else break;
        }
        return closest;
    }
}
